pkg/protocol/commitment.go

```diff
@@ -0,0 +1,285 @@
+}
+
+// deriveIsAboveLatestVerifiedCommitment derives the IsAboveLatestVerifiedCommitment flag of this Commitment which is
+// true if the parent is already above the latest verified Commitment or if the parent is verified and we are not.
+func (c *Commitment) deriveIsAboveLatestVerifiedCommitment(parent *Commitment) func() {
+	return c.IsAboveLatestVerifiedCommitment.DeriveValueFrom(reactive.NewDerivedVariable3(func(_ bool, parentAboveLatestVerifiedCommitment bool, parentIsVerified bool, isVerified bool) bool {
+		return parentAboveLatestVerifiedCommitment || (parentIsVerified && !isVerified)
+	}, parent.IsAboveLatestVerifiedCommitment, parent.IsVerified, c.IsVerified))
+}
+
+// deriveRequestAttestations derives the RequestAttestations flag of this Commitment which is true if our Chain is
+// requesting attestations (while not having an engine), and we are the directly above the latest attested Commitment.
+func (c *Commitment) deriveRequestAttestations(chain *Chain, parent *Commitment) func() {
+	return c.RequestAttestations.DeriveValueFrom(reactive.NewDerivedVariable4(func(_ bool, startEngine bool, verifyAttestations bool, parentIsAttested bool, isAttested bool) bool {
+		return !startEngine && verifyAttestations && parentIsAttested && !isAttested
+	}, chain.StartEngine, chain.RequestAttestations, parent.IsAttested, c.IsAttested))
+}
+
+// deriveWarpSyncBlocks derives the WarpSyncBlocks flag of this Commitment which is true if our Chain is requesting
+// warp sync, and we are the directly above the latest verified Commitment.
+func (c *Commitment) deriveWarpSyncBlocks(chain *Chain, parent *Commitment) func() {
+	return c.WarpSyncBlocks.DeriveValueFrom(reactive.NewDerivedVariable4(func(_ bool, engineInstance *engine.Engine, warpSync bool, parentIsVerified bool, isVerified bool) bool {
+		return engineInstance != nil && warpSync && parentIsVerified && !isVerified
+	}, chain.Engine, chain.WarpSyncMode, parent.IsVerified, c.IsVerified))
+}
+
+// deriveReplayDroppedBlocks derives the ReplayDroppedBlocks flag of this Commitment which is true if our Chain has an
+// engine, is no longer requesting warp sync, and we are above the latest verified Commitment.
+func (c *Commitment) deriveReplayDroppedBlocks(chain *Chain) func() {
+	return c.ReplayDroppedBlocks.DeriveValueFrom(reactive.NewDerivedVariable3(func(_ bool, engineInstance *engine.Engine, warpSyncing bool, isAboveLatestVerifiedCommitment bool) bool {
+		return engineInstance != nil && !warpSyncing && isAboveLatestVerifiedCommitment
+	}, chain.Engine, chain.WarpSyncMode, c.IsAboveLatestVerifiedCommitment))
+}
+
+// forceChain forces the Chain of this Commitment to the given Chain by promoting it to the main child of its parent if
+// the parent is on the target Chain.
+func (c *Commitment) forceChain(targetChain *Chain) {
+	if currentChain := c.Chain.Get(); currentChain != targetChain {
+		if parent := c.Parent.Get(); parent.Chain.Get() == targetChain {
+			parent.MainChild.Set(c)
```

Comment:
Just for my understanding: once we set the `MainChild` it will trigger its updates and via the setup in `deriveChain` we will evict the `targetChain`?

Reply:
yes